File: app/modules/accounts/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import List

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class UsageTrendPoint(DashboardModel):
    t: datetime
    v: float


class AccountUsageTrend(DashboardModel):
    primary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary_scheduled: list[UsageTrendPoint] = Field(default_factory=list)


class AccountUsage(DashboardModel):
    primary_remaining_percent: float | None = None
    secondary_remaining_percent: float | None = None
    monthly_remaining_percent: float | None = None


class AccountRequestUsage(DashboardModel):
    request_count: int = 0
    total_tokens: int = 0
    cached_input_tokens: int = 0
    total_cost_usd: float = 0.0


class AccountTokenStatus(DashboardModel):
    expires_at: datetime | None = None
    state: str | None = None


class AccountAuthStatus(DashboardModel):
    access: AccountTokenStatus | None = None
    refresh: AccountTokenStatus | None = None
    id_token: AccountTokenStatus | None = None


class AccountLimitWarmupStatus(DashboardModel):
    window: str
    reset_at: int
    status: str
    model: str
    attempted_at: datetime
    completed_at: datetime | None = None
    error_code: str | None = None
    error_message: str | None = None


class AccountAdditionalWindow(DashboardModel):
    used_percent: float
    reset_at: int | None = None
    window_minutes: int | None = None


class AccountAdditionalQuota(DashboardModel):
    quota_key: str | None = None
    limit_name: str
    metered_feature: str
    display_label: str | None = None
    routing_policy: str = Field(default="inherit", pattern=r"^(inherit|normal|burn_first|preserve)$")
    primary_window: AccountAdditionalWindow | None = None
    secondary_window: AccountAdditionalWindow | None = None


class AccountSummary(DashboardModel):
    account_id: str
    email: str
    alias: str | None = None
    display_name: str
    workspace_id: str | None = None
    workspace_label: str | None = None
    seat_type: str | None = None
    plan_type: str
    routing_policy: str = Field(default="normal", pattern=r"^(normal|burn_first|preserve)$")
    status: str
    security_work_authorized: bool = False
    usage: AccountUsage | None = None
    reset_at_primary: datetime | None = None
    reset_at_secondary: datetime | None = None
    reset_at_monthly: datetime | None = None
    window_minutes_primary: int | None = None
    window_minutes_secondary: int | None = None
    window_minutes_monthly: int | None = None
    last_refresh_at: datetime | None = None
    capacity_credits_primary: float | None = None
    remaining_credits_primary: float | None = None
    capacity_credits_secondary: float | None = None
    remaining_credits_secondary: float | None = None
    capacity_credits_monthly: float | None = None
    remaining_credits_monthly: float | None = None
    request_usage: AccountRequestUsage | None = None
    additional_quotas: list[AccountAdditionalQuota] = Field(default_factory=list)
    credits_has: bool | None = None
    credits_unlimited: bool | None = None
    credits_balance: float | None = None
    deactivation_reason: str | None = None
    auth: AccountAuthStatus | None = None
    limit_warmup_enabled: bool = False
    limit_warmup: AccountLimitWarmupStatus | None = None
    # True when another account row in the same response shares this real email,
    # ChatGPT account identity, and workspace slot.
    # Operators see this after a token-invalidation cascade where re-adding
    # via OAuth creates a side-by-side row with a fresh refresh token; the
    # older row keeps a revoked token and keeps generating 401s through the
    # load balancer. Flagging the dupes in /accounts lets the dashboard
    # surface a "delete older" action without requiring the operator to
    # group rows by email themselves. See codex-lb #787 (B).
    is_email_duplicate: bool = False


class AccountsResponse(DashboardModel):
    accounts: List[AccountSummary] = Field(default_factory=list)


class AccountImportResponse(DashboardModel):
    account_id: str
    email: str
    workspace_id: str | None = None
    workspace_label: str | None = None
    seat_type: str | None = None
    plan_type: str
    status: str


class OpenCodeOAuthAuth(DashboardModel):
    type: str = "oauth"
    refresh: str
    access: str
    expires: int = Field(ge=0)
    account_id: str | None = None


class OpenCodeAuthJson(DashboardModel):
    openai: OpenCodeOAuthAuth


class AccountOpenCodeAuthExportAccount(DashboardModel):
    account_id: str
    chatgpt_account_id: str | None = None
    email: str


class AccountOpenCodeAuthExportResponse(DashboardModel):
    filename: str
    account: AccountOpenCodeAuthExportAccount
    auth_json: OpenCodeAuthJson


class AccountUpdateRequest(DashboardModel):
    security_work_authorized: bool | None = None


class AccountUpdateResponse(DashboardModel):
    status: str


class AccountPauseResponse(DashboardModel):
    status: str


class AccountReactivateResponse(DashboardModel):
    status: str


class AccountLimitWarmupUpdateRequest(DashboardModel):
    enabled: bool


class AccountLimitWarmupUpdateResponse(DashboardModel):
    status: str
    enabled: bool


class AccountRoutingPolicyUpdateRequest(DashboardModel):
    routing_policy: str = Field(pattern=r"^(normal|burn_first|preserve)$")


class AccountRoutingPolicyUpdateResponse(DashboardModel):
    account_id: str
    routing_policy: str


class AccountDeleteResponse(DashboardModel):
    status: str


class AccountExportResponse(DashboardModel):
    account_id: str
    email: str
    workspace_id: str | None = None
    workspace_label: str | None = None
    seat_type: str | None = None
    plan_type: str
    status: str
    auth_json: str


class AccountProbeRequest(DashboardModel):
    model: str | None = Field(
        default=None,
        description=(
            "Optional model slug for the probe request. Defaults to the service's configured fallback when omitted."
        ),
    )


class AccountProbeResponse(DashboardModel):
    status: str
    account_id: str
    probe_status_code: int
    primary_used_percent_before: float | None = None
    primary_used_percent_after: float | None = None
    secondary_used_percent_before: float | None = None
    secondary_used_percent_after: float | None = None
    account_status_before: str
    account_status_after: str


class AccountTrendsResponse(DashboardModel):
    account_id: str
    primary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary_scheduled: list[UsageTrendPoint] = Field(default_factory=list)


class CodexAuthTokens(DashboardModel):
    id_token: str = Field(serialization_alias="id_token", validation_alias="id_token")
    access_token: str = Field(serialization_alias="access_token", validation_alias="access_token")
    refresh_token: str = Field(serialization_alias="refresh_token", validation_alias="refresh_token")
    account_id: str | None = Field(
        default=None,
        serialization_alias="account_id",
        validation_alias="account_id",
    )


class CodexAuthJson(DashboardModel):
    auth_mode: str = Field(default="chatgpt", serialization_alias="auth_mode", validation_alias="auth_mode")
    openai_api_key: str | None = Field(
        default=None,
        serialization_alias="OPENAI_API_KEY",
        validation_alias="OPENAI_API_KEY",
    )
    tokens: CodexAuthTokens
    last_refresh: str = Field(serialization_alias="last_refresh", validation_alias="last_refresh")


class AccountAuthExportTokens(DashboardModel):
    id_token: str
    access_token: str
    refresh_token: str
    expires_at_ms: int = Field(ge=0)


class AccountAuthExportResponse(DashboardModel):
    filename: str
    account: AccountOpenCodeAuthExportAccount
    tokens: AccountAuthExportTokens
    codex_auth_json: CodexAuthJson
    opencode_auth_json: OpenCodeAuthJson


class AccountAliasRequest(DashboardModel):
    alias: str | None = Field(default=None, max_length=255)


class AccountAliasResponse(DashboardModel):
    account_id: str
    alias: str | None = None

File: frontend/src/features/accounts/schemas.ts
import { z } from "zod";

export const UsageTrendPointSchema = z.object({
  t: z.string().datetime({ offset: true }),
  v: z.number(),
});

export const AccountUsageTrendSchema = z.object({
  primary: z.array(UsageTrendPointSchema),
  secondary: z.array(UsageTrendPointSchema),
  secondaryScheduled: z.array(UsageTrendPointSchema).default([]),
});

export const AccountUsageSchema = z.object({
  primaryRemainingPercent: z.number().nullable(),
  secondaryRemainingPercent: z.number().nullable(),
});

export const AccountRequestUsageSchema = z.object({
  requestCount: z.number().int().nonnegative(),
  totalTokens: z.number().int().nonnegative(),
  cachedInputTokens: z.number().int().nonnegative(),
  totalCostUsd: z.number().nonnegative(),
});

export const AccountTokenStatusSchema = z.object({
  expiresAt: z.string().datetime({ offset: true }).nullable().optional(),
  state: z.string().nullable().optional(),
});

export const AccountAuthSchema = z.object({
  access: AccountTokenStatusSchema.nullable().optional(),
  refresh: AccountTokenStatusSchema.nullable().optional(),
  idToken: AccountTokenStatusSchema.nullable().optional(),
});

export const AccountLimitWarmupStatusSchema = z.object({
  window: z.enum(["primary", "secondary"]).or(z.string()),
  resetAt: z.number().int(),
  status: z.string(),
  model: z.string(),
  attemptedAt: z.string().datetime({ offset: true }),
  completedAt: z.string().datetime({ offset: true }).nullable().optional(),
  errorCode: z.string().nullable().optional(),
  errorMessage: z.string().nullable().optional(),
});

export const AccountAdditionalWindowSchema = z.object({
  usedPercent: z.number(),
  resetAt: z.number().nullable().optional(),
  windowMinutes: z.number().nullable().optional(),
});

export const AccountAdditionalQuotaSchema = z.object({
  quotaKey: z.string().nullable().optional(),
  limitName: z.string(),
  meteredFeature: z.string(),
  displayLabel: z.string().nullable().optional(),
  primaryWindow: AccountAdditionalWindowSchema.nullable().optional(),
  secondaryWindow: AccountAdditionalWindowSchema.nullable().optional(),
});

export const AccountSummarySchema = z.object({
  accountId: z.string(),
  email: z.string(),
  alias: z.string().nullable().optional(),
  displayName: z.string(),
  workspaceId: z.string().nullable().optional(),
  workspaceLabel: z.string().nullable().optional(),
  seatType: z.string().nullable().optional(),
  planType: z.string(),
  status: z.string(),
  usage: AccountUsageSchema.nullable().optional(),
  resetAtPrimary: z.string().datetime({ offset: true }).nullable().optional(),
  resetAtSecondary: z.string().datetime({ offset: true }).nullable().optional(),
  windowMinutesPrimary: z.number().nullable().optional(),
  windowMinutesSecondary: z.number().nullable().optional(),
  capacityCreditsSecondary: z.number().nullable().optional(),
  remainingCreditsSecondary: z.number().nullable().optional(),
  requestUsage: AccountRequestUsageSchema.nullable().optional(),
  auth: AccountAuthSchema.nullable().optional(),
  additionalQuotas: z.array(AccountAdditionalQuotaSchema).default([]),
  limitWarmupEnabled: z.boolean().default(false),
  limitWarmup: AccountLimitWarmupStatusSchema.nullable().optional(),
  isEmailDuplicate: z.boolean().optional(),
});

export const AccountTrendsResponseSchema = z.object({
  accountId: z.string(),
  primary: z.array(UsageTrendPointSchema),
  secondary: z.array(UsageTrendPointSchema),
  secondaryScheduled: z.array(UsageTrendPointSchema).default([]),
});

export const AccountsResponseSchema = z.object({
  accounts: z.array(AccountSummarySchema),
});

export const AccountImportResponseSchema = z.object({
  accountId: z.string(),
  email: z.string(),
  workspaceId: z.string().nullable().optional(),
  workspaceLabel: z.string().nullable().optional(),
  seatType: z.string().nullable().optional(),
  planType: z.string(),
  status: z.string(),
});

export const OpenCodeOAuthAuthSchema = z.object({
  type: z.literal("oauth"),
  refresh: z.string(),
  access: z.string(),
  expires: z.number().int().nonnegative(),
  accountId: z.string().nullable().optional(),
});

export const OpenCodeAuthJsonSchema = z.object({
  openai: OpenCodeOAuthAuthSchema,
});

export const AccountOpenCodeAuthExportAccountSchema = z.object({
  accountId: z.string(),
  chatgptAccountId: z.string().nullable().optional(),
  email: z.string(),
});

export const CodexAuthTokensSchema = z.object({
  id_token: z.string(),
  access_token: z.string(),
  refresh_token: z.string(),
  account_id: z.string().nullable().optional(),
});

export const CodexAuthJsonSchema = z.object({
  auth_mode: z.string(),
  OPENAI_API_KEY: z.string().nullable().optional(),
  tokens: CodexAuthTokensSchema,
  last_refresh: z.string(),
});

export const AccountAuthExportTokensSchema = z.object({
  idToken: z.string(),
  accessToken: z.string(),
  refreshToken: z.string(),
  expiresAtMs: z.number().int().nonnegative(),
});

export const AccountAuthExportResponseSchema = z.object({
  filename: z.string(),
  account: AccountOpenCodeAuthExportAccountSchema,
  tokens: AccountAuthExportTokensSchema,
  codexAuthJson: CodexAuthJsonSchema,
  opencodeAuthJson: OpenCodeAuthJsonSchema,
});

export const AccountActionResponseSchema = z.object({
  status: z.string(),
});

export const AccountAliasRequestSchema = z.object({
  alias: z.string().max(255).nullable(),
});

export const AccountAliasResponseSchema = z.object({
  accountId: z.string(),
  alias: z.string().nullable(),
});

export const AccountLimitWarmupUpdateRequestSchema = z.object({
  enabled: z.boolean(),
});

export const AccountLimitWarmupUpdateResponseSchema = z.object({
  status: z.string(),
  enabled: z.boolean(),
});

export const OauthStartRequestSchema = z.object({
  forceMethod: z.string().optional(),
});

export const OauthStartResponseSchema = z.object({
  flowId: z.string().nullable().optional(),
  method: z.string(),
  authorizationUrl: z.string().nullable(),
  callbackUrl: z.string().nullable(),
  verificationUrl: z.string().nullable(),
  userCode: z.string().nullable(),
  deviceAuthId: z.string().nullable(),
  intervalSeconds: z.number().nullable(),
  expiresInSeconds: z.number().nullable(),
});

export const OauthStatusResponseSchema = z.object({
  status: z.string(),
  errorMessage: z.string().nullable(),
});

export const OauthCompleteRequestSchema = z.object({
  flowId: z.string().optional(),
  deviceAuthId: z.string().optional(),
  userCode: z.string().optional(),
});

export const OauthCompleteResponseSchema = z.object({
  status: z.string(),
});

export const ManualOauthCallbackRequestSchema = z.object({
  callbackUrl: z.string(),
  flowId: z.string().optional(),
});

export const ManualOauthCallbackResponseSchema = z.object({
  status: z.string(),
  errorMessage: z.string().nullable(),
});

export const RuntimeConnectAddressResponseSchema = z.object({
  connectAddress: z.string(),
});

export const OAuthStateSchema = z.object({
  flowId: z.string().nullable().optional(),
  status: z.enum(["idle", "starting", "pending", "success", "error"]),
  method: z.enum(["browser", "device"]).nullable(),
  authorizationUrl: z.string().nullable(),
  callbackUrl: z.string().nullable(),
  verificationUrl: z.string().nullable(),
  userCode: z.string().nullable(),
  deviceAuthId: z.string().nullable(),
  intervalSeconds: z.number().nullable(),
  expiresInSeconds: z.number().nullable(),
  errorMessage: z.string().nullable(),
});

export const ImportStateSchema = z.object({
  status: z.enum(["idle", "uploading", "success", "error"]),
  message: z.string().nullable(),
});

export type UsageTrendPoint = z.infer<typeof UsageTrendPointSchema>;
export type AccountUsageTrend = z.infer<typeof AccountUsageTrendSchema>;
export type AccountSummary = z.infer<typeof AccountSummarySchema>;
export type AccountAliasResponse = z.infer<typeof AccountAliasResponseSchema>;
export type AccountLimitWarmupStatus = z.infer<typeof AccountLimitWarmupStatusSchema>;
export type AccountAdditionalQuota = z.infer<typeof AccountAdditionalQuotaSchema>;
export type AccountTrendsResponse = z.infer<typeof AccountTrendsResponseSchema>;
export type OpenCodeAuthJson = z.infer<typeof OpenCodeAuthJsonSchema>;
export type CodexAuthJson = z.infer<typeof CodexAuthJsonSchema>;
export type AccountAuthExportTokens = z.infer<typeof AccountAuthExportTokensSchema>;
export type AccountAuthExportResponse = z.infer<typeof AccountAuthExportResponseSchema>;
export type OauthStartResponse = z.infer<typeof OauthStartResponseSchema>;
export type OauthStatusResponse = z.infer<typeof OauthStatusResponseSchema>;
export type ManualOauthCallbackResponse = z.infer<typeof ManualOauthCallbackResponseSchema>;
export type RuntimeConnectAddressResponse = z.infer<
  typeof RuntimeConnectAddressResponseSchema
>;
export type OAuthState = z.infer<typeof OAuthStateSchema>;
export type ImportState = z.infer<typeof ImportStateSchema>;
